### ISSUE TITLE
First VVP prototype made: what do you think?

### ISSUE THREAD

**Issue by djgroen** [djgroen/FabSim3#141]
I developed a first VVP that performs a validation function on an ensemble of simulation outputs, and then uses an aggregation function to combine the validation outcomes into a compound metric.

The VVP is quite short and simple, and I have been able to leverage it directly using the FabFlee `validate_flee` function.

You can find the VVP here:
https://github.com/djgroen/FabSim3/blob/master/VVP/vvp.py

I am very curious what you all think of this initial design, and am happy to take any suggestions to improve on it.

**Comment by wedeling** [comment]
@djgroen
I played around with `vvp.py` a bit, and I think it could be useful. I made a few suggestions but the main structure remains the same, see

https://github.com/wedeling/FabUQCampaign/blob/master/examples/ocean_2D_paper/vvp.py

Just plotting the samples from one (or more) EasyVVUQ campaigns could be a useful verification task, which I can do with `vvp.py` via

```
    #make a histrogram from the samples for each EasyVVUQ campaign
    vvp.ensemble_vvp([my_campaign_p4.campaign_dir + '/runs', 
                      my_campaign_p5.campaign_dir + '/runs',
                      my_campaign_p6.campaign_dir + '/runs'], 
                      load_uq_csv_output, plt.hist, 'E_mean')
```

Printing / plotting the post-processed statistical moments or Sobol indices (useful for a convergence study) can also be done:

```
    #print the results for all EasyVVUQ campaigns found in the work directory
    vvp.ensemble_vvp(work_dir, load_uq_results, print)
```

Things like `load_uq_results` are now part of my script, but probably should become part of EasyVVUQ sampler objects.

**Comment by djgroen** [comment]
@wedeling I like your suggestions as a whole, although there is one point that I think needs more discussion:

"2) 'validation_function': same comment, changed it to 'partition_function',
   being the opposite of 'aggregation_function', something that acts on a single sample only"

The name partition_function implies that the function performs a partitioning, whereas it is my understanding that the function performs V&V-related tasks on individual samples from the simulation, and does neither a partition or an aggregation. Should we perhaps rename *partition_function* to *sample_testing_function* perhaps, or can you think of an even better name?

- [ ] Incorporate @wedeling's changes to FabSim VVP definition.
- [ ] Agree to sensible function names in this first VVP implementation.
- [ ] Wire up revised VVP implementation to the migration use case.

**Comment by wedeling** [comment]
@djgroen Yes, I wasn't sure about that one. 'sample_testing_function' is fine by me

**Comment by djgroen** [comment]
Okay, I made the changes accordingly and update the documentation. Do we have our first VVP in place then, or would you like to make further changes? :)

**Comment by wedeling** [comment]
@derek. I guess we have our 1st one. Note sure though if this will become part of EasyVVUQ or of a separate VV package

**Comment by djgroen** [comment]
In my opinion, VVPs should be entirely library-independent files.  So, once we have a second one, I think we should make a separate repository for them :).

Hopefully, we'll find these VVP codes in all sorts of places in- and outside the VECMA toolkit in due time :).

**Comment by djgroen** [comment]
@wedeling There's one more idea that came up while struggling to sleep due to jet-lag ;).

I fully understand that we removed the `print()` statements in the initial version of the VVP, but thinking deeper about it, I do think it is a bit hacky to do any print/exporting work purely as part of the `sample_test_function` and the `aggregation_function`.

I'm therefore thinking of establishing an optional extra argument, which is for a `representation_function`, which could look like this:

```
def representation_function(sample_test_function_outputs, aggregation_function_output):
```

This would then be implemented as follows:

- If `representation_function = None`, skip and don't do any exporting / printing.
- If `representation_function = print`, dump the output using a trivial print statement.
- Else, use the function provided to e.g., (a) write selected aspects of the function outputs to screen or (b) to write them to one or more files. The exact output specification would be determined by the representation_function (e.g., it could also produce an image if that is preferable by the application developer).

What do you think? Do you think such a convention would strengthen this VVP, and potentially VVPs in general?

P.s., the signature of our VVP would then of course become:
`def ensemble_vvp(results_dirs, sample_testing_function, aggregation_function, representation_function=None, **kwargs):`

**Comment by djgroen** [comment]
We made good progress on this for M18, but I'm leaving this open for M21 because of the last item.
